httpMiddleware.go

```diff
@@ -0,0 +1,199 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/palantir/go-githubapp/githubapp"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/rs/zerolog"
+)
+
+type statusCodeResponseWriter struct {
+	http.ResponseWriter
+	statusCode int
+}
+
+func (w *statusCodeResponseWriter) WriteHeader(code int) {
+	w.statusCode = code
+	w.ResponseWriter.WriteHeader(code)
+}
+func InitializeResponder(ctx context.Context) context.Context {
+	var responder func(http.ResponseWriter, *http.Request)
+	return context.WithValue(ctx, responderKey{}, &responder)
+}
+
+type responderKey struct{}
+type LoggerFunc func(msg string, m map[string]interface{})
+
+func addContextMiddleware(logger zerolog.Logger, h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		//eventType := r.Header.Get("X-GitHub-Event")
+		//deliveryID := r.Header.Get("X-GitHub-Delivery")
+
+		ctx := logger.With()
+		//ctx = ctx.Str("github_event_type", eventType)
+		//ctx = ctx.Str("github_delivery_id", deliveryID)
+
+		l := ctx.Logger()
+
+		r = r.WithContext(l.WithContext(r.Context()))
+
+		h.ServeHTTP(w, r)
+	})
+}
+
+func loggerMiddleware(logger LoggerFunc, h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		statusWriter := &statusCodeResponseWriter{w, http.StatusOK}
+		start := time.Now()
+		h.ServeHTTP(statusWriter, r)
+
+		// request duration in miliseconds
+		duration := time.Since(start).Nanoseconds() / 1e6
+		statusCode := statusWriter.statusCode
+
+		logMessage := fmt.Sprintf("[%d] %s %s", statusCode, r.Method, r.URL.String())
+		logMap := map[string]interface{}{
+			"responseTime": fmt.Sprintf("%d", duration),
+		}
+
+		logger(logMessage, logMap)
+	})
+}
+
+func inboundMetricsMiddleware(promRegisterer prometheus.Registerer, h http.Handler) http.Handler {
+
+	httpRequests := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "inbound_http_status_code_total",
+			Help: "Counter of HTTP status codes of inbound HTTP requests",
+		},
+		[]string{"status_code", "method"},
+	)
+	promRegisterer.MustRegister(httpRequests)
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		statusWriter := &statusCodeResponseWriter{w, http.StatusOK}
+		//start := time.Now()
+		h.ServeHTTP(statusWriter, r)
+
+		// request duration in miliseconds
+		//duration := time.Since(start).Nanoseconds() / 1e6
+		statusCode := statusWriter.statusCode
+
+		httpRequests.WithLabelValues(strconv.Itoa(statusCode), r.Method).Inc()
+
+	})
+}
+
+type roundTripperFunc func(*http.Request) (*http.Response, error)
+type ClientMiddleware func(http.RoundTripper) http.RoundTripper
```

Comment:
This type is not used as far as I can tell?

Reply:
Correct, it's because I though I could use my own type. But it was required I used `githubapp.ClientMiddleware` in 

`func githubMetricsMiddleware(promRegisterer prometheus.Registerer) githubapp.ClientMiddleware {`

Reply:
https://github.com/lunarway/release-manager-bot/pull/21/commits/1cf2aa5032ff85ec32656127329667f1c8f1eb78